Support `sw.os` and `sw.kernel` contract feature labels

cywang117 opened the pull request:
Relates-to: https://github.com/balena-os/balena-supervisor/pull/2444
Relates-to: https://github.com/balena-io/docs/pull/3259
Change-type: minor

pipex reviewed (commented):
I don't really have comments on the code. I don't think I looked at the spec before it got bet on and that's on me. I think both sw.kernel and sw.os should validate both slug and version. 


So the `sw.kernel` contract should be 
```json
{
  "type": "sw.kernel",
  "slug": "linux",
  "version": ">5.4"
}
```

And `sw.os`

```json
{
  "type": "sw.os",
  "slug": "balena-os",
  "version": ">=6"
}
```

With that, the labels should be 

`io.balena.features.requires.sw.linux: <version>` and `io.balena.features.requires.sw.balena-os: <version>`

I made a comment on this on the discussion thread but I failed to comment on the spec

cywang117 committed:
Update contracts in fixtures to use 'balena-os' slug for 'sw.os' type

Also run lint-fix

Signed-off-by: Christina Ying Wang <christina@balena.io>

cywang117 committed:
Support sw.os and sw.kernel contract requirement variants, allow multiple variants

This allows users to specify an OS slug as a contract requirement, for example
io.balena.features.requires.sw.balena-os:>VERSION, which transforms to
{type:'sw.os',slug:'balena-os',version:'>VERSION'}. Similarly,
io.balena.features.requires.sw.linux:>VERSION transforms to
{type:'sw.kernel',slug:'linux',version:'>VERSION'}. This allows for
eventual flexibility in OS and kernel requirements for a container.

Currently we only support the 'balena-os' sw.os variant and the
'linux' sw.kernel variant.

See: https://balena.fibery.io/Work/Project/Add-kernel-and-OS-version-to-container-contracts-1916
Change-type: minor
Signed-off-by: Christina Ying Wang <christina@balena.io>

pipex commented on lib/parse/compose.ts:
```diff
@@ -557,9 +598,23 @@ function validateLabels(labels: Dict<string>) {
+			if (contract in contractParser) {
+				contractParser[contract].validate(value, name);
+
+				// Check for duplicate contract types (e.g., multiple sw.os or sw.kernel)
```
Why? Couldn't the same container support balena-os and ubuntu?

cywang117 replied:
You're right, I removed the restriction on one sw.os/sw.kernel variant in `validateLabels`. Now, if the user specifies multiple sw.os or sw.kernel variants, provided we support them, the labels generated from these labels will be combined under an `or` clause. See test [here](https://github.com/balena-io-modules/balena-compose/pull/117/changes#diff-93d4920157e11c832ff9c7439b1042735c3488aa83075c8b235ff03bcc5dafd8R355-R468)

pipex commented on lib/parse/compose.ts:
```diff
@@ -362,7 +362,7 @@ function normalizeServiceVolumes(
 		switch (ref.type) {
 			case 'bind':
-				mounts.push(ref.source!);
+				mounts.push(ref.source ?? '');
```
I don't like this, while technically the right side should never happen, the newer line of code has a different connotation. I think this can be solved by more clever typing of `VolumeRef`, something like 

```ts
type VolumeRef = { type: 'bind', /* bind properties */ } | { type: 'tmpfs', /* ... */ } | { /* ... * }
```

If possible try to see if that works.

cywang117 replied:
This TS syntax error is actually made irrelevant with the integration of [balena-compose-parser](https://github.com/balena-io-modules/balena-compose-parser/blob/main/lib/compose.ts) into balena-compose in https://github.com/balena-io-modules/balena-compose/pull/97, which will hopefully be merged soon. Originally I changed it in this PR to satisfy the linter for testing. I don't feel a strong need to go over this code that will be replaced soon so I'll just revert it to what it was before, as it now seems to not be raising a linter error for some reason, probably because I ran `npm i` under a different Node version
